rasterio/_io.pyx

```diff
@@ -828,17 +828,41 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+def is_zip(data):
+    """Detect zip compressed data"""
+    return data[:4].startswith(b'PK\x03\x04')
+
+
 cdef class MemoryFileBase(object):
     """Base for a BytesIO-like class backed by an in-memory file."""
 
-    def __init__(self, initial_bytes=b''):
-        """Map bytes to a file in an in-memory filesystem."""
+    def __init__(self, file_or_bytes=None):
+        """A file in an in-memory filesystem.
+
+        Parameters
+        ----------
+        file_or_bytes : file or bytes
+            A file opened in binary mode or bytes or a bytearray
```

Comment:
`BytesIO` for example doesn't have a `mode` attribute but I'll look for a way to fail early and clearly.